js/ui/appIconBar.js

```diff
@@ -810,21 +812,59 @@ const ScrolledIconList = new Lang.Class({
         });
         newChild.connect('app-icon-unpinned', () => {
             favorites.removeFavorite(app.get_id());
-            if (app.state == Shell.AppState.STOPPED) {
-                newActor.destroy();
-                this._taskbarApps.delete(app);
-                this._updatePage();
-            }
         });
         this._taskbarApps.set(app, newChild);
 
-        this._container.add_actor(newActor);
+        if (oldChild) {
+            this._container.replace_child(oldChild.actor, newActor);
+        } else {
+            this._container.add_actor(newActor);
+        }
```

Comment:
Nitpick: does not need the curly braces

Reply:
Ah yes, this style is even spelled out in HACKING! Will fix.